proposal.html

```diff
@@ -1,40 +1,37 @@
 <pre class=metadata>
-title: Array.prototype.flatMap &amp; Array.prototype.flatten
+title: Array.prototype.flatMap
 toc: false
 stage: 3
 contributors: Michael Ficarra and Brian Terlson
 </pre>
 
 <emu-intro id="intro">
   <h1>Introduction</h1>
-  <p>`Array.prototype.flatten` returns a new array with all sub-array elements concatted into it recursively up to the specified depth.</p>
-  <p>`Array.prototype.flatMap` first maps each element using a mapping function, then flattens the result into a new array. It is identical to a map followed by a flatten of depth 1, but flatMap is quite often useful and merging both into one method is slightly more efficient.</p>
+  <p>`Array.prototype.flatMap` returns a new array with all sub-array elements (up to a depth limit) recursively added into it, optionally first mapping each element using a function. It is identical to a map followed by a flatten, but flatMap is quite often useful and merging both into one method is slightly more efficient.</p>
 </emu-intro>
 
 <emu-clause id="sec-Array.prototype.flatMap">
-  <h1>Array.prototype.flatMap ( _mapperFunction_ [ , _thisArg_ ] )</h1>
-  <p>When the `flatMap` method is called with one or two arguments, the following steps are taken:</p>
-  <emu-alg>
-    1. Let _O_ be ? ToObject(*this* value).
-    1. Let _sourceLen_ be ? ToLength(? Get(_O_, `"length"`)).
-    1. If IsCallable(_mapperFunction_) is *false*, throw a *TypeError* exception.
-    1. If _thisArg_ is present, let _T_ be _thisArg_; else let _T_ be *undefined*.
-    1. Let _A_ be ? ArraySpeciesCreate(_O_, 0).
-    1. Perform ? FlattenIntoArray(_A_, _O_, _sourceLen_, 0, 1, _mapperFunction_, _T_).
-    1. Return _A_.
-  </emu-alg>
-</emu-clause>
-<emu-clause id="sec-Array.prototype.flatten">
-  <h1>Array.prototype.flatten( [ _depth_ ] )</h1>
-  <p>When the `flatten` method is called with zero or one arguments, the following steps are taken:</p>
+  <h1>Array.prototype.flatMap ( [ _depth_ ] [ , _mapperFunction_ [ , _thisArg_ ] ] )</h1>
+  <p>When the `flatMap` method is called with one or two or three arguments, the following steps are taken:</p>
   <emu-alg>
     1. Let _O_ be ? ToObject(*this* value).
     1. Let _sourceLen_ be ? ToLength(? Get(_O_, `"length"`)).
     1. Let _depthNum_ be 1.
+    1. Let _T_ be *undefined*.
+    1. If IsCallable(_depth_) is *true*, then
+      1. Set _thisArg_ to _mapperFunction_.
```

Comment:
This kind of argument switching is unfortunately common in userland, but i don’t think there’s any precedent for it in the language.

What would the .length be of .flatMap if all three arguments are optional? I’d expect a length of 1, at least.

Reply:
Length should be zero, just like `flatten` is/was.
```js
[ [0, "foo"], [1, "bar"] ].flatMap()	// => [ 0, "foo", 1, "bar" ]
```

Reply:
ah, gotcha. In that case then i think overloading one function to have two purposes is an unwise api design.

Reply:
"One function with two purposes" is the entire [_premise_](https://github.com/tc39/proposal-flatMap/blob/master/proposal.html#L11) of `flatMap`:
> `Array.prototype.flatMap` first maps each element using a mapping function, then flattens the result into a new array. It is identical to a map followed by a flatten of depth 1, but flatMap is quite often useful and merging both into one method is slightly more efficient.